### PR TITLE
[Modules] : New module Microsoft.EventGrid domains

### PR DESCRIPTION
**Pull request by shaban-khan**
# Description

I have added new module for Microsoft.EventGrid domains: https://learn.microsoft.com/en-us/azure/templates/microsoft.eventgrid/domains?pivots=deployment-language-bicep

This module is under Microsoft.EventGrid and manage large number of Event Grid topics related to the same application. It allows an event publisher to publish events to thousands of topics at the same time. Domains also give you authentication and authorization control over each topic so you can partition your tenants.

# Pipeline references
> For module/pipeline changes, please create and attach the status badge of your successful run.

|  Pipeline  |

[EventGrid: Domains](https://github.com/shaban-khan/ResourceModules/actions/workflows/ms.eventgrid.domains.yml)

# Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [ ] I did format my code

### CODE REVIEW

**Comment by shaban-khan** [comment]
checklist updated

File: settings.yml [hunk 1]
```diff
@@ -68,7 +68,7 @@ variables:
   # ------------------------------- #
 
   bicepRegistryDoPublish: true # Set to true, if you would like to publish module templates to a bicep registry
-  bicepRegistryName: adpsxxazacrx001 # The name of the bicep registry (ACR) to publish to. If it does not exist, it will be created.
+  bicepRegistryName: skhanadpsxxazacrx001 # The name of the bicep registry (ACR) to publish to. If it does not exist, it will be created.
```
Comment (AlexanderSehr) [inline settings.yml:71]: Please undo the changes in this file as they are specific to your environment. 

In my experience the 'easiest' way to contribute is to open a PR from a branch so that you can rollback the changes in the settings.yml without needing to change it in your main branch.

File: modules/Microsoft.EventGrid/domains/.bicep/nested_roleAssignments.bicep [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  'Monitoring Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '749f88d5-cbae-40b8-bcfc-e573ddc772fa')
+  'Monitoring Reader': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '43d0d8ad-25c7-4714-9337-8ba259a9fe05')
+  Owner: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8e3af657-a8ff-443c-a75c-2fe8c4bcb635')
+  Reader: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
+  'Resource Policy Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '36243c78-bf99-498c-9df9-86d9f8d28608')
+  'Role Based Access Control Administrator (Preview)': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'f58310d9-a9f6-439a-9e8d-f62e7b41a168')
+  'User Access Administrator': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '18d7d88d-d35e-4fb5-a5c3-7773c20a72d9')
+}
+
+resource eventGrid 'Microsoft.EventGrid/domains@2022-06-15' existing = {
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.bicep/nested_roleAssignments.bicep:58]:
```suggestion
resource domain 'Microsoft.EventGrid/domains@2022-06-15' existing = {
```

File: modules/Microsoft.EventGrid/domains/.bicep/nested_roleAssignments.bicep [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  properties: {
+    description: description
+    roleDefinitionId: contains(builtInRoleNames, roleDefinitionIdOrName) ? builtInRoleNames[roleDefinitionIdOrName] : roleDefinitionIdOrName
+    principalId: principalId
+    principalType: !empty(principalType) ? any(principalType) : null
+    condition: !empty(condition) ? condition : null
+    conditionVersion: !empty(conditionVersion) && !empty(condition) ? conditionVersion : null
+    delegatedManagedIdentityResourceId: !empty(delegatedManagedIdentityResourceId) ? delegatedManagedIdentityResourceId : null
+  }
+  scope: eventGrid
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.bicep/nested_roleAssignments.bicep:73]:
```suggestion
  scope: domain
```

File: modules/Microsoft.EventGrid/domains/.bicep/nested_roleAssignments.bicep [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  'Role Based Access Control Administrator (Preview)': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'f58310d9-a9f6-439a-9e8d-f62e7b41a168')
+  'User Access Administrator': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '18d7d88d-d35e-4fb5-a5c3-7773c20a72d9')
+}
+
+resource eventGrid 'Microsoft.EventGrid/domains@2022-06-15' existing = {
+  name: last(split(resourceId, '/'))
+}
+
+resource roleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = [for principalId in principalIds: {
+  name: guid(eventGrid.id, principalId, roleDefinitionIdOrName)
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.bicep/nested_roleAssignments.bicep:63]:
```suggestion
  name: guid(domain.id, principalId, roleDefinitionIdOrName)
```

File: modules/Microsoft.EventGrid/domains/.test/min/deploy.test.bicep [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+// ========== //
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.eventgrid.domains-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'egtmin'
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.test/min/deploy.test.bicep:14]:
```suggestion
param serviceShort string = 'egdmin'
```

File: modules/Microsoft.EventGrid/domains/.test/common/deploy.test.bicep [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+// ========== //
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.eventgrid.domains-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'egtcom'
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.test/common/deploy.test.bicep:14]:
```suggestion
param serviceShort string = 'egdcom'
```

File: modules/Microsoft.EventGrid/domains/.test/pe/deploy.test.bicep [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+// ========== //
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.eventgrid.domains-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'egtpe'
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.test/pe/deploy.test.bicep:14]:
```suggestion
param serviceShort string = 'egdpe'
```

File: modules/Microsoft.EventGrid/domains/.test/pe/dependencies.bicep [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+@description('Required. The name of the Virtual Network to create.')
+param virtualNetworkName string
+
+resource virtualNetwork 'Microsoft.Network/virtualNetworks@2022-01-01' = {
+    name: virtualNetworkName
+    location: location
+    properties: {
+        addressSpace: {
+            addressPrefixes: [
+                '10.0.0.0/24'
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.test/pe/dependencies.bicep:13]:
```suggestion
                addressPrefix
```

File: modules/Microsoft.EventGrid/domains/.test/pe/dependencies.bicep [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            ]
+        }
+        subnets: [
+            {
+                name: 'defaultSubnet'
+                properties: {
+                    addressPrefix: '10.0.0.0/24'
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.test/pe/dependencies.bicep:20]:
```suggestion
                    addressPrefix: addressPrefix
```

File: modules/Microsoft.EventGrid/domains/.test/pe/dependencies.bicep [hunk 10]
```diff
@@ -0,0 +1,47 @@
+@description('Optional. The location to deploy to.')
+param location string = resourceGroup().location
+
+@description('Required. The name of the Virtual Network to create.')
+param virtualNetworkName string
+
+resource virtualNetwork 'Microsoft.Network/virtualNetworks@2022-01-01' = {
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.test/pe/dependencies.bicep:7]:
```suggestion
param virtualNetworkName string

var addressPrefix = '10.0.0.0/24'

resource virtualNetwork 'Microsoft.Network/virtualNetworks@2022-01-01' = {
```

File: modules/Microsoft.EventGrid/domains/.test/common/dependencies.bicep [hunk 11]
```diff
@@ -0,0 +1,58 @@
+@description('Optional. The location to deploy to.')
+param location string = resourceGroup().location
+
+@description('Required. The name of the Virtual Network to create.')
+param virtualNetworkName string
+
+@description('Required. The name of the Managed Identity to create.')
+param managedIdentityName string
+
+resource virtualNetwork 'Microsoft.Network/virtualNetworks@2022-01-01' = {
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.test/common/dependencies.bicep:10]:
```suggestion
param managedIdentityName string

var addressPrefix = '10.0.0.0/24'

resource virtualNetwork 'Microsoft.Network/virtualNetworks@2022-01-01' = {
```

File: modules/Microsoft.EventGrid/domains/.test/common/dependencies.bicep [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    name: virtualNetworkName
+    location: location
+    properties: {
+        addressSpace: {
+            addressPrefixes: [
+                '10.0.0.0/24'
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.test/common/dependencies.bicep:16]:
```suggestion
                addressPrefix
```

File: modules/Microsoft.EventGrid/domains/.test/common/dependencies.bicep [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+            ]
+        }
+        subnets: [
+            {
+                name: 'defaultSubnet'
+                properties: {
+                    addressPrefix: '10.0.0.0/24'
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/.test/common/dependencies.bicep:23]:
```suggestion
                    addressPrefix: addressPrefix
```

File: modules/Microsoft.EventGrid/domains/readme.md [hunk 14]
```diff
@@ -0,0 +1,524 @@
+# EventGrid Domains `[Microsoft.EventGrid/domains]`
+
+This module deploys EventGrid Domains.
+// TODO: Replace Resource and fill in description
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/readme.md:4]:
```suggestion
```

File: modules/Microsoft.EventGrid/domains/readme.md [hunk 15]
```diff
@@ -0,0 +1,524 @@
+# EventGrid Domains `[Microsoft.EventGrid/domains]`
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/readme.md:1]: Please update the file with the used `Set-ModuleReadMe` function once we resolved all other comments (as they will likely have an effect on the file)

File: modules/Microsoft.EventGrid/domains/readme.md [hunk 16]
```diff
@@ -0,0 +1,524 @@
...
+| `privateEndpoints` | array | `[]` |  | Configuration details for private endpoints. For security reasons, it is recommended to use private endpoints whenever possible. |
+| `publicNetworkAccess` | string | `''` | `['', Disabled, Enabled]` | Whether or not public network access is allowed for this resource. For security reasons it should be disabled. If not specified, it will be disabled by default if private endpoints are set and inboundIpRules are not set. |
+| `roleAssignments` | array | `[]` |  | Array of role assignment objects that contain the 'roleDefinitionIdOrName' and 'principalId' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: '/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11'. |
+| `tags` | object | `{object}` |  | Tags of the resource. |
+
+
+### Parameter Usage: `<ParameterPlaceholder>`
+
+// TODO: Fill in Parameter usage
+
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/readme.md:60]:
```suggestion
```

File: modules/Microsoft.EventGrid/domains/deploy.bicep [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+  }
+}
+
+resource domain 'Microsoft.EventGrid/domains@2022-06-15' = {
+  name: name
+  location: location
+  tags: tags
+  properties: {
+    publicNetworkAccess: !empty(publicNetworkAccess) ? any(publicNetworkAccess) : (!empty(privateEndpoints) && empty(inboundIpRules) ? 'Disabled' : null)
+    inboundIpRules: (empty(inboundIpRules) ? null : inboundIpRules)
```
Comment (AlexanderSehr) [inline modules/Microsoft.EventGrid/domains/deploy.bicep:132]:
```suggestion
    inboundIpRules: !empty(inboundIpRules) ? inboundIpRules : null
```

File: settings.yml (hunk 1 above)
Comment (shaban-khan) [reply settings.yml]: reverted changes in the settings.yml file as per suggestion

File: modules/Microsoft.EventGrid/domains/readme.md (hunk 15 above)
Comment (shaban-khan) [reply modules/Microsoft.EventGrid/domains/readme.md:1]: Successfully ran function Set-ModuleReadMe to updated the readme.md file. 

The module workflow run successfully except Publishing due to the ACR name in settings.yml file

**Comment by shaban-khan** [comment]
> @shaban-khan please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree [company="{Microsoft}"]

**Comment by shaban-khan** [comment]
> @shaban-khan the command you issued was incorrect. Please try again.
> 
> Examples are:
> 
> ```
> @microsoft-github-policy-service agree
> ```
> 
> and
> 
> ```
> @microsoft-github-policy-service agree company="your company"
> ```

**Comment by shaban-khan** [comment]
@microsoft-github-policy-service agree

and

@microsoft-github-policy-service agree company="Microsoft"

**Review by AlexanderSehr — APPROVED** [review]
